tools/hsmtool.c

```diff
@@ -75,161 +92,71 @@ static bool ensure_hsm_secret_exists(int fd, const char *path)
+		printf("Enter hsm_secret password:\n");
+		fflush(stdout);
+		passphrase = read_stdin_pass(tmpctx, &err);
+		if (!passphrase)
+			errx(EXITCODE_ERROR_HSM_FILE, "Could not read password: %s", hsm_secret_error_str(err));
+	}
 
-	exit_code = hsm_secret_encryption_key_with_exitcode(passwd, &key, &err);
-	if (exit_code > 0)
-		errx(exit_code, "%s", err);
-	if (!decrypt_hsm_secret(&key, &encrypted_secret, hsm_secret))
-		errx(ERROR_LIBSODIUM, "Could not retrieve the seed. Wrong password ?");
+	hsms = extract_hsm_secret(ctx, contents, tal_bytelen(contents), passphrase, &err);
+	if (!hsms) {
+		errx(EXITCODE_ERROR_HSM_FILE, "%s", hsm_secret_error_str(err));
+	}
+	return hsms;
 }
 
-/* Taken from hsmd. */
-static void get_channel_seed(struct secret *channel_seed, struct node_id *peer_id,
-                             u64 dbid, struct secret *hsm_secret)
+/* Legacy function - only works with binary encrypted format */
+static int decrypt_hsm(const char *hsm_secret_path)
 {
-	struct secret channel_base;
-	u8 input[sizeof(peer_id->k) + sizeof(dbid)];
-	/*~ Again, "per-peer" should be "per-channel", but Hysterical Raisins */
-	const char *info = "per-peer seed";
+	int fd;
+	struct hsm_secret *hsms;
+	const char *dir, *backup;
 
-	/*~ We use the DER encoding of the pubkey, because it's platform
-	 * independent.  Since the dbid is unique, however, it's completely
-	 * unnecessary, but again, existing users can't be broken. */
-	/* FIXME: lnd has a nicer BIP32 method for deriving secrets which we
-	 * should migrate to. */
-	hkdf_sha256(&channel_base, sizeof(struct secret), NULL, 0,
-	            hsm_secret, sizeof(*hsm_secret),
-	            /*~ Initially, we didn't support multiple channels per
-	             * peer at all: a channel had to be completely forgotten
-	             * before another could exist.  That was slightly relaxed,
-	             * but the phrase "peer seed" is wired into the seed
-	             * generation here, so we need to keep it that way for
-	             * existing clients, rather than using "channel seed". */
-	             "peer seed", strlen("peer seed"));
-	memcpy(input, peer_id->k, sizeof(peer_id->k));
-	BUILD_ASSERT(sizeof(peer_id->k) == PUBKEY_CMPR_LEN);
-	/*~ For all that talk about platform-independence, note that this
-	 * field is endian-dependent!  But let's face it, little-endian won.
-	 * In related news, we don't support EBCDIC or middle-endian. */
-	memcpy(input + PUBKEY_CMPR_LEN, &dbid, sizeof(dbid));
+	/* Check if it's a format we can decrypt */
+	u8 *contents = grab_file(tmpctx, hsm_secret_path);
+	if (!contents)
+		errx(EXITCODE_ERROR_HSM_FILE, "Reading hsm_secret");
 
-	hkdf_sha256(channel_seed, sizeof(*channel_seed),
-	            input, sizeof(input),
-	            &channel_base, sizeof(channel_base),
-	            info, strlen(info));
-}
+	tal_resize(&contents, tal_bytelen(contents) - 1);
+	enum hsm_secret_type type = detect_hsm_secret_type(contents, tal_bytelen(contents));
```

Comment:
Again, you don't need to resize, you can just hand tal_bytelen(contents)-1 to detect_hsm_secret_type...